crypto/fipsmodule/rand/ctrdrbg.c

```diff
@@ -12,21 +12,205 @@
+
+  for (size_t block = 0; block < input_len; block += AES_BLOCK_SIZE) {
+    CRYPTO_xor16(output, output, &input[block]);
+    AES_encrypt(output, output, key);
+  }
+  return 1;
+}
+
+// CTR_DRBG_df implements the block cipher derivation function from SP800-90A
+// 10.3.2 assuming a constant keylen of 256, block size of 16, and output length
+// of |CTR_DRBG_ENTROPY_LEN|. Arguments |entropy_input| and may alias
+// |entropy_output| but not partially overlap.
+// Returns 1 on success, 0 otherwise.
+static int CTR_DRBG_df(CTR_DRBG_STATE *drbg,
+                uint8_t *entropy_input, size_t entropy_input_len,
+                uint8_t entropy_output[CTR_DRBG_ENTROPY_LEN]) {
+
+  int ret = 0;
+
+  if (drbg == NULL ||
+      entropy_input == NULL ||
+      entropy_output == NULL ||
+      entropy_input_len == 0) {
+    return 0;
+  }
+
+  // 10.3.2 (8) is constant when key length is constant. Initialise an AES
+  // object once which can be used read-only by all threads. 
+  CRYPTO_once(bcc_init_once_bss_get(), init_df_bcc_once);
+
+  // Generally, integer values for 10.3.2 must be representable by a
+  // 32-bit integer. First condition is to capture cases were the size_t type
+  // is smaller than 0xFFFFFFFF e.g. 32-bit Android and the compiler throwing
+  // warnings.
+  if (SIZE_MAX > UINT32_MAX && entropy_input_len > UINT32_MAX) {
+    return 0;
+  }
+
+  // The derivation function can be split into two phases:
+  // A) Run BCC to extract.
+  // B) Run block cipher iteratively to expand.
+  //
+  // We first handle phase (A). The goal is to construct strings IV and S, that
+  // are used as input to BCC. IV and S are composed of the following values:
+  // i, L, N (32-bit represented)
+  // 10.3.2 (2): L = entropy_input_len
+  // 10.3.2 (3): N = CTR_DRBG_ENTROPY_LEN
+  // 10.3.2 (4+5): S = L || N || input_string || 0x80 || 0x00-padded to block size
+  // 10.3.2 (9.1): IV = i || 0x00-padded to block size
+  //
+  // Using these values, we construct IV_S = IV || S used in 10.3.2 (9.2).
+
+  static const size_t df_s_prefix_length = 8; // 2x 32-bit integers
+  // Compute required zero-padding (might be zero). +1 because of 0x80
+  // delimiter in S.
+  size_t df_s_zero_padding = (-(df_s_prefix_length + entropy_input_len + 1)) & (AES_BLOCK_SIZE-1);
+
+  // Knowing the padding, we can compute the full length of S.
+  size_t df_s_length = df_s_prefix_length + entropy_input_len + 1 + df_s_zero_padding;
+
+  // Since the length of the IV is fixed when the output length is fixed, we can
+  // compute the length of IV || S.
+  static const size_t df_iv_length = 16;
+  size_t df_iv_s_length = df_iv_length + df_s_length;
+
+  // Zeroise to avoid zero padding later.
+  uint8_t *IV_S = OPENSSL_zalloc(df_iv_s_length);
+  if (IV_S == NULL) {
+    return 0;
+  }
+
+  // IV_S = IV || L || N || entropy_input
+  CRYPTO_store_u32_be(IV_S + df_iv_length, (uint32_t) entropy_input_len);
+  CRYPTO_store_u32_be(IV_S + df_iv_length + 4, (uint32_t) CTR_DRBG_ENTROPY_LEN); 
```

Comment:
```
CRYPTO_store_u32_be(IV_S + df_iv_length + 4, (uint32_t) CTR_DRBG_ENTROPY_LEN); 
```

I understand it's `4` since `entropy_input_len` is a 4 byte value, but feels off to have a "magic number" here. Perhaps `sizeof(uint32_t)` would be sufficient?

Or perhaps having constants that represent the offsets where values begin rather then the length of a value? Multiple ways to do it I guess.